proposals/0672-declarative-xcode-projects.md

```diff
@@ -0,0 +1,92 @@
+---
+title: A declarative approach to Xcode projects
+author:
+- Riccardo Cipolleschi
+date: 2023-06-19
+---
+
+# RFC0672: A declarative approach to Xcode projects
+
+## Summary
+
+Managing Xcodeproj files is error prone and time consuming as Xcode generates random UUIDs to identify entities in the project.
+
+Historically, we had trouble with updates, especially when we were modifying the scripts in the Xcode build script phases: those changes are buried in a wall of text that the `Xcodeproj` is and they are easy to miss.
+
+It also uses a proprietary syntax: it is not a XML, a JSON or anything standard. Preexistent tools can't really highlight changes in a meaningful way, making the Xcodeproj files just a wall of test which can feel daunting for the final user.
+
+The current proposal suggests to move away from manually managing `Xcodeproj` files in favor of a more declarative description. There are different tools in the OSS space that allows to describe the project declaratively and that then reads that description and generate a valid Xcodeproj accordingly.
+
+The suggested too is [Tuist.io](https://tuist.io): it allows to describe the project in Swift, leveraging type checking, syntax highlights and code completion. It also uses the default language for Apple development, which is familiar to all the iOS developers and it is executable code, providing more flexibility than a pure markup language.
+
+## Basic example
+
+You can have a look at it in action in this PR:
+
+- https://github.com/facebook/react-native/pull/37952
+
+The idea is to describe the project with a declarative approach using Swift. Tuist allows you to load the `Project.swift` file in an instance of Xcode throught the `tuist edit` command. This allows to leverage code completion, type checking and syntax highlight to ensure the correctness of the project.
+
+## Motivation
+
+Managing Xcodeproj files is error prone and time consuming. Those files are source of frustrations for teams using them and they have been reported as one of the main pain point when users have to migrate to a new version of React Native.
```

Comment:
Can you clarify the meaning of "managing Xcodeproj files" for someone who (like me) doesn't have a lot of iOS-specific context? What does this involve in practice?

Reply:
Currently, project files are committed in the repository.

For how Xcode works, every time you modify a settings,  add a file to your project or change a script, the project files get modified.

It is super common for teams to work at the same time on a project, adding files to implement features which in turn generates some changes to the Xcodeproject. Xcode uses randomly generated UUID to refer to entities in the project and the format is a proprietary one and it is also very verbose.

The common, general pain point is conflict resolution:
- Team  A adds some files
- Team B adds some other files
- Xcode puts those files, with their UUIDs in the same file location.
- That generates a git conflict on several lines which seems unrelated that need to be handled manually
- It is super easy to make a mistake and resulting in an invalid project.

The second category of problems are specific to React Native, when users have to update their project:
- Post Build script are plain texts and buried into the verbosity of the project. Several times, people makes mistakes forgetting to update those scripts.
- Changes in build settings are buried into the project and are hard to spot. Xcode has 2 way to define build settings (directly in the Xcode proj and in some xconfig file... which unfortunately are baked into the project as well) and usually there are 2 config (Debug and Release) which ends up for updater in having to check 4 configurations to make sure not to lose any update
- Renaming a file HAS to happen within Xcode. Xcode does not have a mechanism to crawl the file system, so if you rename your `AppDelegate.m` to `AppDelegate.mm` from a terminal, Xcode will not see that change. The declarative tool can `glob` on folders to avoid this problem, creating a project using all the files in that folder.